### PR TITLE
Add FirewallRule class

### PR DESCRIPTION
**Pull request by ndowmon**

### CODE REVIEW

File: src/schemas/FirewallRule.json [hunk 1]
```diff
@@ -0,0 +1,59 @@
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
```
Comment (mknoedel) [inline src/schemas/FirewallRule.json:2]: Will we need to port these changes to the AWS integration? It looks like we just put these rules in an array on a key for that integration.
https://bitbucket.org/jupiterone/jupiter-integration-aws/src/b1b082a76600f85738911adbea9d4112869059e7/src/integration-aws/ec2/converters/entities.ts?at=master#entities.ts-583

Comment (mknoedel) [reply src/schemas/FirewallRule.json:2]: I also assume you already took a look at [Rule](https://github.com/JupiterOne/data-model/blob/master/src/schemas/Rule.json), but thought I'd bring that up as well.

**Review by mknoedel — DISMISSED** [review]
Looks good to me so long as you have determined we need this. I would like a little documentation on the PR for easier debugging in the future before you merge this though.

File: src/schemas/FirewallRule.json (hunk 1 above)
Comment (ndowmon) [reply src/schemas/FirewallRule.json:2]: Yeah, if this goes through I think it'd be good to get it into existing integrations like AWS, since the `JSON.stringify(ipPermission)` operation makes these properties useless from a query language perspective.

I have looked at `Rule` but IMO a firewall rule is a super-specific class of configuration (as indicated by the common properties shared across all firewall rules)

File: src/schemas/FirewallRule.json [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  "allOf": [
+    { "$ref": "#Entity" },
+    {
+      "properties": {
+        "priority": {
+          "description": "Priority of the firewall rule. Typically, a number between 0 and 65535.",
+          "type": "integer"
+        },
+        "source": {
+          "description": "Source of the firewall rule. Can be IP address, CIDR range, 'Any', or other firewall-defined options.",
```
Comment (austinkelleher) [inline src/schemas/FirewallRule.json:15]: Should we recommend only `*` instead of `Any`?

File: src/schemas/FirewallRule.json [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          "type": "string"
+        },
+        "sourceIp": {
+          "description": "Source IP of the firewall rule.",
+          "type": "string",
+          "format": "ip"
+        },
+        "sourcePort": {
+          "description": "Source port of the firewall rule. Typically an integer between 0 and 65535, but could also be 'Any', '*', or range of ports.",
```
Comment (austinkelleher) [inline src/schemas/FirewallRule.json:24]: `Any`, `*`, or "range of ports", will not be accepted when the`type` is `integer`

File: src/schemas/FirewallRule.json [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+          "description": "Destination of the firewall rule. Can be IP address, CIDR range, 'Any', or other firewall-defined options.",
+          "type": "string"
+        },
+        "destinationIp": {
+          "description": "Destination IP of the firewall rule.",
+          "type": "string",
+          "format": "ip"
+        },
+        "destinationPort": {
+          "description": "Destination port of the firewall rule. Typically an integer between 0 and 65535, but could also be 'Any', '*', or range of ports.",
```
Comment (austinkelleher) [inline src/schemas/FirewallRule.json:37]: Should we recommend only `*` instead of `Any`?

File: src/schemas/FirewallRule.json [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+          "type": "integer"
+        },
+        "protocol": {
+          "description": "The protocol ",
+          "type": "string",
+          "examples": ["TCP", "UDP", "Any"]
```
Comment (austinkelleher) [inline src/schemas/FirewallRule.json:43]: Same question about recommending `*` here?

**Review by austinkelleher — CHANGES_REQUESTED** [review]

**Comment by erkangz** [comment]
I did not follow the change here closely. Why is this class needed? 

I think we have a Rule and RuleSet class already?

**Comment by erkangz** [comment]
With the AWS integration, details of each rule are captures as properties on the relationship between the entity and what is allowed to/from the entity. 

I'm concerned that adding this will unnecessarily add more entities, which is undesired from both our perspective (data volume) and customer's (cost).

**Comment by ndowmon** [comment]
@erkangz yes, I see how this works in AWS. My concern with putting firewall rules on edges is that we cannot be as expressive on edges as we can on nodes (at least in J1QL). For example, there's not currently a way in J1QL to express direction of edges, representing ingress/egress, and edges do not allow array properties for IP addresses, ports, or protocols defined by these rules. Those are some of my main concerns when assigning firewall rules to edges. 

As for creating a new class `FirewallRule` as opposed to using `Rule`, this is a pretty rare case where almost all firewall rules will share this collection of properties (`sourcePort`, `sourceIp`, `destinationPort`, `destinationIp`, `protocol`, `action`, `direction`), and none of these properties would be valid for almost any other type of `Rule`, so the distinction in my mind lends `FirewallRule` to its own classification.

**Comment by erkangz** [comment]
> @erkangz yes, I see how this works in AWS. My concern with putting firewall rules on edges is that we cannot be as expressive on edges as we can on nodes (at least in J1QL). For example, there's not currently a way in J1QL to express direction of edges, representing ingress/egress, and edges do not allow array properties for IP addresses, ports, or protocols defined by these rules. Those are some of my main concerns when assigning firewall rules to edges.
> 
> As for creating a new class `FirewallRule` as opposed to using `Rule`, this is a pretty rare case where almost all firewall rules will share this collection of properties (`sourcePort`, `sourceIp`, `destinationPort`, `destinationIp`, `protocol`, `action`, `direction`), and none of these properties would be valid for almost any other type of `Rule`, so the distinction in my mind lends `FirewallRule` to its own classification.

Valid concern. However, that's a query lang issue/limitation and I don't think we should change the data-model because of that. For direction, specifically to firewall rules, we do have `ingress` and `egress` boolean properties on the edge that can be used to filter on direction.

**Comment by aiwilliams** [comment]
tldr; I would stick with the current approach until we can prove it won't work (perhaps examples from Azure or Google Cloud) and we should develop docs describing the firewall model and how to implement it.

A few thoughts:

* We can make any entity class non-billable.
* There would be `number of rules X 2` more objects if we had another entity between a `Firewall` and the resources it relates to because of a rule (`Firewall -> FirewallRule -> (Host|IpAddress|Network)` vs `Firewall -> (Host|IpAddress|Network)`), but I think it very little data in the big scheme of things.
* We want to relate the `Firewall` to each `(Host|IpAddress|Network)` represented in a rule, so it isn't necessary for these rule relationships to support `string[]` properties.
* It is reasonable to anticipate we can make the query language support relationships better, though I believe we've added properties to entities in the past to get them into the index and duplicated property names to make queries look better, efforts that acknowledge it could be some time before the query language/service supports features to avoid those ingestion decisions.
* It is understandable that early decisions can make it difficult to change things now, and there would need to be good reasons to go through pains to do so.

Here are some reasons I recall that relationships were chosen.

1. Firewall rules simply define relationships between the `Firewall` and `(Host|IpAddress|Network)` entities, similar to a policy document (rules could be added as raw data of a `Firewall`).
2. This query reads nicely: `find Host that PROTECTS Firewall that ALLOWS Internet`.
3. This query reads less nice: `find Host that PROTECTS Firewall that HAS Rule that ALLOWS Internet`.
4. This query allows listing specific rules just fine: `find Host that PROTECTS Firewall that ALLOWS as rule Network where rule.inbound=true return rule.*`.

@erkangz, do you have others?

**Comment by aiwilliams** [comment]
Note that there is a TypeScript type for rule relationships: https://github.com/JupiterOne/data-model/blob/de4d2ff81d31fa3f07a6c578ce11bd0a71518bbe/src/relationships.ts#L1
